### PR TITLE
[Frontend] Add experimental flag to skip non-inlinable function bodies

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch adds an experimental flag to skip non-inlinable function
bodies and turns it on for the standard library and overlay
`.swiftmodules`.

This flag is disabled if the requested action generates IR.

---

This option causes some small improvements in the time it takes to type check and generate an optimized and unoptimized standard library `.swiftmodule`

| Mode                                 | Optimization | Time     | Time (Skipping) | Ratio |
|:-------------------------------------|:-------------|:---------|:----------------|:------|
| `-emit-parseable-module-interface`   | N/A          | 16.254s  | 14.947s         | 1.09x |
| `-emit-module`                       | `-Onone`     | 21.685s  | 19.462s         | 1.11x |
| `-emit-module`                       | `-O`         | 50.526s  | 35.234s         | 1.43x |

However, since the standard library is full of inlinable code, it doesn't benefit much from skipping non-inlinable code.

To find a more representative Swift project with little inlinable code, I ran these benchmarks against the current master of SwiftLint, which shows much stronger wins:

| Mode                                 | Optimization | Time     | Time (Skipping) | Ratio     |
|:-------------------------------------|:-------------|:---------|:----------------|:----------|
| `-emit-parseable-module-interface`   | N/A          | 13.179s  | 1.732s          | **7.70x** |
| `-emit-module`                       | `-Onone`     | 16.21s   | 2.719s          | **5.88x** |
| `-emit-module`                       | `-O`         | 63.408s  | 12.186s         | **5.26x** |

### CODE REVIEW

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
Okay, looks like I’m missing some specializations or some inlinable code.

**Comment by harlanhaskins** [comment]
The Linux failures might shed some light on what I'm missing:

```
Test Result (4 failures / +3)
  - Swift(linux-x86_64).SILOptimizer.static_arrays.swift
  - Swift(linux-x86_64).SILOptimizer.optionset.swift
  - Swift(linux-x86_64).SILOptimizer.pointer_conversion.swift
  - Swift(linux-x86_64).SILOptimizer.array_contentof_opt.swift
```

**Comment by harlanhaskins** [comment]
Oops! Was skipping inlinable deinitializers.

@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
Still missing _something_...

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test linux platform

**Comment by harlanhaskins** [comment]
Running these again given the @inline(__always) changes.

@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
Just gonna kick off another benchmarking run with @jrose-apple's recent pointer-related changes in #21126.

@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
Seems that wasn't it.

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
Gotta remember to leave SwiftOnoneSupport alone...

@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
Getting much closer!

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -425,7 +425,10 @@ class alignas(1 << DeclAlignInBits) Decl {
     Synthesized : 1,
 
     /// Whether this member's body consists of a single expression.
-    HasSingleExpressionBody : 1
+    HasSingleExpressionBody : 1,
+
+    /// Whether the function body was skipped during typechecking.
+    WasBodyTypecheckingSkipped : 1
```
Comment (slavapestov) [inline include/swift/AST/Decl.h]: The existing BodyKind should capture this already

File: lib/AST/DeclContext.cpp [hunk 2]
```diff
@@ -340,10 +340,16 @@ swift::ResilienceExpansionRequest::evaluate(Evaluator &evaluator,
       return ResilienceExpansion::Maximal;
     }
 
-    // Stored property initializer contexts use minimal resilience expansion
-    // if the type is formally fixed layout.
-    if (isa<PatternBindingInitializer>(dc)) {
-      if (auto *NTD = dyn_cast<NominalTypeDecl>(dc->getParent())) {
+    // Non-static stored property initializer contexts use minimal resilience
+    // expansion if the type is formally fixed layout.
```
Comment (slavapestov) [inline lib/AST/DeclContext.cpp]: This looks like an unrelated change, it warrants its own commit and tests

File: lib/Sema/TypeCheckStmt.cpp [hunk 3]
```diff
@@ -2139,13 +2139,25 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
...
 
+  // If we were asked to skip non-inlinable bodies, and this function is
+  // non-inlinable, then skip typechecking the body. As a conservative step,
+  // still do typecheck functions that appear inside a local context, because
+  // their parent context will have been typechecked.
+  if (tc.SkipNonInlinableFunctionBodies &&
+      !AFD->getDeclContext()->isLocalContext() &&
+      AFD->getResilienceExpansion() == ResilienceExpansion::Maximal) {
+    AFD->setBodyTypecheckingSkipped(true);
+    return false;
```
Comment (slavapestov) [inline lib/Sema/TypeCheckStmt.cpp]: It would be better if we just didn't add these functions to TypeChecker::definedFunctions. We shouldn't get to running this request on them at all

File: lib/Sema/TypeCheckError.cpp [hunk 4]
```diff
@@ -1627,6 +1627,10 @@ void TypeChecker::checkFunctionErrorHandling(AbstractFunctionDecl *fn) {
   // by the time we got here.
   if (!fn->hasInterfaceType()) return;
 
+  // If we skipped typechecking the body, don't perform this check; it will
+  // always warn.
+  if (fn->wasBodyTypecheckingSkipped()) return;
```
Comment (slavapestov) [inline lib/Sema/TypeCheckError.cpp]: Would be better if checkFunctionErrorHandling() didn't get called on these functions at all

File: lib/Sema/TypeCheckStmt.cpp [hunk 5]
```diff
@@ -2139,13 +2139,25 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
...
   BraceStmt *body = AFD->getBody();
   if (!body || AFD->isBodyTypeChecked())
     return false;
 
+  // If we were asked to skip non-inlinable bodies, and this function is
+  // non-inlinable, then skip typechecking the body. As a conservative step,
+  // still do typecheck functions that appear inside a local context, because
+  // their parent context will have been typechecked.
+  if (tc.SkipNonInlinableFunctionBodies &&
+      !AFD->getDeclContext()->isLocalContext() &&
```
Comment (slavapestov) [inline lib/Sema/TypeCheckStmt.cpp]: Why is there an exception for local functions?

File: lib/Serialization/Serialization.cpp [hunk 6]
```diff
@@ -4909,6 +4909,12 @@ void Serializer::writeAST(ModuleOrSourceFile DC,
     for (auto TD : localTypeDecls) {
       hasLocalTypes = true;
       Mangle::ASTMangler Mangler;
+
+      // If we skipped the parent of this type, don't bother serializing it.
+      if (auto afd = dyn_cast<AbstractFunctionDecl>(TD->getDeclContext()))
```
Comment (slavapestov) [inline lib/Serialization/Serialization.cpp]: A local type can appear inside other local contexts, such as a property initializer or a closure

File: lib/Sema/TypeChecker.cpp [hunk 7]
```diff
@@ -400,6 +400,13 @@ void swift::performTypeChecking(SourceFile &SF, TopLevelContext &TLC,
     if (Options.contains(TypeCheckingFlags::ForImmediateMode))
       TC.setInImmediateMode(true);
 
+    if (Options.contains(TypeCheckingFlags::SkipNonInlinableFunctionBodies))
+      // Disable this optimization if we're compiling SwiftOnoneSupport, because
```
Comment (slavapestov) [inline lib/Sema/TypeChecker.cpp]: We shouldn't be generating a swiftinterface for SwiftOnoneSupport, though?

File: lib/SILGen/SILGen.cpp [hunk 8]
```diff
@@ -803,10 +803,13 @@ emitMarkFunctionEscapeForTopLevelCodeGlobals(SILLocation loc,
     TopLevelSGF->B.createMarkFunctionEscape(loc, Captures);
 }
 
-void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
+bool SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
   // Emit any default argument generators.
   emitDefaultArgGenerators(AFD, AFD->getParameters());
 
+  if (AFD->wasBodyTypecheckingSkipped())
```
Comment (slavapestov) [inline lib/SILGen/SILGen.cpp]: Instead, I would prefer it if the skipping was done further up the callstack. Returning a boolean is subtle

**Review by slavapestov — COMMENTED** [review]
Cool! You should investigate enabling delayed parsing for primary files in this mode as well, so that we don't even have to build AST for their bodies.

**Comment by slavapestov** [comment]
Even more things to investigate:

- There are parts of SILGenType you can look into skipping as well, for example emitting serialized vtables and witness tables.
- Skipping global variables that do not have inlinable accessors.
- To check your work: add a SIL pass that asserts that when the flag is set, there aren't any non-serializable SILFunctions in the final SILModule.

File: lib/SILGen/SILGen.cpp [hunk 9]
```diff
@@ -803,10 +803,13 @@ emitMarkFunctionEscapeForTopLevelCodeGlobals(SILLocation loc,
     TopLevelSGF->B.createMarkFunctionEscape(loc, Captures);
 }
 
-void SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
+bool SILGenModule::emitAbstractFuncDecl(AbstractFunctionDecl *AFD) {
   // Emit any default argument generators.
   emitDefaultArgGenerators(AFD, AFD->getParameters());
```
Comment (slavapestov) [inline lib/SILGen/SILGen.cpp]: Will this still emit default arg initializers for private/internal functions?

File: lib/SILGen/SILGen.cpp (hunk 8 above)
Comment (slavapestov) [reply lib/SILGen/SILGen.cpp]: ... or further down. See how emitOrDelayFunction takes a SILDeclRef and a callback? If the SILDeclRef's isSerialized() is not set, you can just not call the callback and return. This will handle default args, accessors, functions, and most other things, except for vtable and witness thunks.

File: include/swift/AST/Decl.h [hunk 10]
```diff
@@ -5714,14 +5714,17 @@ class AbstractFunctionDecl : public GenericContext, public ValueDecl {
   /// Note that the body was skipped for this function.  Function body
   /// cannot be attached after this call.
   void setBodySkipped(SourceRange bodyRange) {
-    assert(getBodyKind() == BodyKind::None);
+//    assert(getBodyKind() == BodyKind::None ||
```
Comment (harlanhaskins) [inline include/swift/AST/Decl.h]: This needs a FIXME, I'd like to change the assert to the commented version once we enable delayed parsing everywhere.

File: lib/Sema/TypeChecker.cpp (hunk 7 above)
Comment (harlanhaskins) [reply lib/Sema/TypeChecker.cpp]: This also applies to -emit-module

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
This benchmark isn't actually testing the feature, I just want to make sure I didn't regress anything when the flag isn't set.

@swift-ci please benchmark

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
Python lint failure

@swift-ci please smoke test macOS

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test macOS

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

File: lib/SILGen/SILGen.cpp [hunk 11]
```diff
@@ -1644,6 +1644,22 @@ class SourceFileScope {
 
 } // end anonymous namespace
 
+// Determines if the provided type decl is embedded at some point in an
+// AbstractFunctionDecl whose body has been skipped.
+// FIXME: Duplicated from lib/Serialization/Serialization.cpp
+static bool isInSkippedFunctionContext(const TypeDecl *td) {
```
Comment (CodaFi) [inline lib/SILGen/SILGen.cpp]: This might be good to lift into libAST

File: lib/Sema/TypeCheckStmt.cpp [hunk 12]
```diff
@@ -1835,82 +1835,6 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
   return BS;
 }
 
-static Optional<unsigned>
-getParamIndex(const ParameterList *paramList, const ParamDecl *decl) {
```
Comment (CodaFi) [inline lib/Sema/TypeCheckStmt.cpp]: Can this change be made in isolation?

File: lib/Sema/TypeCheckStmt.cpp [hunk 13]
```diff
@@ -2145,7 +2069,6 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
 
   // FIXME(InterfaceTypeRequest): Remove this.
   (void)AFD->getInterfaceType();
```
Comment (CodaFi) [inline lib/Sema/TypeCheckStmt.cpp]: You should be able to get rid of this too.

**Comment by harlanhaskins** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckStmt.cpp (hunk 13 above)
Comment (harlanhaskins) [reply lib/Sema/TypeCheckStmt.cpp]: Tried it, there are a couple of code completion issues, likely a missing forced interface type. Leaving for a future PR.

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
@swift-ci please test

File: include/swift/Option/Options.td [hunk 14]
```diff
@@ -254,6 +254,10 @@ def stats_output_dir: Separate<["-"], "stats-output-dir">,
 def trace_stats_events: Flag<["-"], "trace-stats-events">,
   Flags<[FrontendOption, HelpHidden]>,
   HelpText<"Trace changes to stats in -stats-output-dir">;
+def experimental_skip_non_inlinable_function_bodies:
+  Flag<["-"], "experimental-skip-non-inlinable-function-bodies">,
+  Flags<[FrontendOption, HelpHidden]>,
+  HelpText<"Skip type-checking and SIL generation for non-inlinable function bodies">;
```
Comment (jrose-apple) [inline include/swift/Option/Options.td]: Nitpick: why is this in the middle of all the stats options?

(Also, why is it a Driver option again?)

File: lib/SILOptimizer/UtilityPasses/NonInlinableFunctionSkippingChecker.cpp [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+  // Thunks and specializations are automatically synthesized, so it's fine that
+  // they end up in the module.
+  if (F.isThunk() || F.isSpecialization())
+    return false;
+
+  // FIXME: We can probably skip property initializers, too.
+  auto func = F.getLocation().getAsASTNode<AbstractFunctionDecl>();
+  if (!func)
+    return false;
```
Comment (jrose-apple) [inline lib/SILOptimizer/UtilityPasses/NonInlinableFunctionSkippingChecker.cpp]: Yeah, this scares me. All sorts of synthesized things will slip through here too.

File: lib/SILOptimizer/UtilityPasses/NonInlinableFunctionSkippingChecker.cpp [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+  // If a body is synthesized/implicit, it shouldn't be skipped.
+  if (func->isImplicit())
+    return false;
+
+  // Local function bodies in inlinable code are okay to show up in the module.
+  if (func->getDeclContext()->isLocalContext())
+    return false;
```
Comment (jrose-apple) [inline lib/SILOptimizer/UtilityPasses/NonInlinableFunctionSkippingChecker.cpp]: Worth checking that the enclosing context is in fact inlinable.

File: lib/Sema/TypeChecker.h [hunk 17]
```diff
@@ -707,6 +711,14 @@ class TypeChecker final : public LazyResolver {
     SwitchCheckingInvocationThreshold = invocationCount;
   }
 
+  void setSkipNonInlinableBodies(bool skip) {
+    SkipNonInlinableFunctionBodies = skip;
+  }
+
+  bool canSkipNonInlinableBodies() const {
+    return SkipNonInlinableFunctionBodies;
+  }
```
Comment (jrose-apple) [inline lib/Sema/TypeChecker.h]: Nitpick: why does this need a getter and setter?

File: test/Frontend/skip-non-inlinable-function-bodies.swift [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+// 1. Make sure you can't -emit-ir or -c when you're skipping non-inlinable function bodies
+
+// RUN: not %target-swift-frontend -emit-ir %s -experimental-skip-non-inlinable-function-bodies %s 2>&1 | %FileCheck %s --check-prefix ERROR
+// RUN: not %target-swift-frontend -c %s -experimental-skip-non-inlinable-function-bodies %s 2>&1 | %FileCheck %s --check-prefix ERROR
+// ERROR: -experimental-skip-non-inlinable-function-bodies does not support emitting IR
+
+// 2. Emit the SIL for a module and check it against the expected strings in function bodies.
+//    If we're doing the right skipping, then the strings that are CHECK-NOT'd won't have been typechecked or SILGen'd.
+
+// RUN: %target-swift-frontend -emit-sil -O -experimental-skip-non-inlinable-function-bodies %s 2>&1 | %FileCheck %s --check-prefixes CHECK,CHECK-SIL-ONLY
```
Comment (jrose-apple) [inline test/Frontend/skip-non-inlinable-function-bodies.swift]: Looks like you didn't end up using CHECK-SIL-ONLY.

File: test/Frontend/skip-non-inlinable-function-bodies.swift [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+
+// 3. Emit the module interface while skipping non-inlinable function bodies. Check it against the same set of strings.
+
+// RUN: %target-swift-frontend -typecheck %s -enable-library-evolution -emit-module-interface-path %t/Module.skipping.swiftinterface -experimental-skip-non-inlinable-function-bodies
+// RUN: %FileCheck %s --check-prefixes CHECK,CHECK-INTERFACE-ONLY < %t/Module.skipping.swiftinterface
```
Comment (jrose-apple) [inline test/Frontend/skip-non-inlinable-function-bodies.swift]: …or CHECK-INTERFACE-ONLY
